docs/proposals/predictive-horizontal-pod-autoscaling.md

```diff
@@ -0,0 +1,232 @@
+changes to record them.
+
+#### Recording `AveragePodInitializationTime`
+
+`AveragePodInitializeTime` has a couple of important nuances. First, it is not
+simply the amount of time between when a pod is ordered to be created and when
+it is created. Rather, we want to record the amount of time from when a pod is
+ordered to be created and when it is ready to perform the task for which it is
+created. For example, a pod may be created, but if the purpose of the pod is to
+run a webserver, and that web server takes five minutes to be able to serve
+requests, then the mere fact that the pod is created does not help balance the
+load nor does it accomplish the purpose of autoscaling. It is the difference
+between the pod’s creation time and the time at which the pod is ready to
+perform its task that we are interested in; we call this metric
+`InitializationTime`.
+
+We will calculate `AveragePodInitializationTime` by storing a measure of
+`InitializationTime` for each pod created during auto-scaling in the pod’s
+`Annotations`. We can then use the pod selector specified by the horizontal pod
+auto-scaler to retrieve and average these values, resulting in the
+`AverageInitializationTime` metric.
+
+We ensure each newly created pod tracks `InitializationTime` in
+`ObjectMeta.Annotations` by adding (in
+`pkg/controller/podautoscaler/horizontal.go`) a watcher for any modifications
+to pods that are created by the horizontal autoscaler. If the modification
+event for these pods indicates that they are now “ready”, then we set the
+`InitializationTime` to be the difference between the time at which they became
+ready and the time at which the pod was ordered to be created.
+
+It would then be possible to calculate the `AveragePodInitializationTime` for pods
+in a horizontal controller group by listing all pods created by this autoscaler
+and averaging their `InitializationTime`.
+
+#### Recording `CPUUtilizationDerivative`
+
+Our proposed algorithm for predictive scaling requires us to know the
+derivative of the metric we are using to determine scaling behavior, in this
+case `CPUUtilizationDerivative`. Thus, we will need to modify the
+`HorizontalPodAutoscalerStatus` struct in `pkg/apis/extensions/v1beta1/types.go` to record one extra field that will support the performance of a simple derivative calculation.
+
+```go
+type HorizontalPodAutoscalerStatus struct {
+...
+
+    // current average CPU utilization over all pods...
+    CurrentCPUUtilizationPercentage *int `json:”currentCPUUtilizationPercentage ...”`
+
+    // previous average CPU utilization over all pods…
+    PreviousCPUUtilizationPercentage *int `json:”previousCPUUtilizationPercentage…”`
+
+}
+```
+
+Recording this field would require the following general modifications to
+record the `CurrentCPUUtilizationPercentage` in the
+`PreviousCPUUtilizationPercentage` field before updating
+`CurrentCPUUtilizationPercentage`.
```

Comment:
How are you going to determine what is the time difference between the previous and the current? We shouldn't rely on hpa control loop timer. We will need to last update time to status (either directly or by annotation).

Moreover, I think we do not need to keep previous - we can just compare previous value of current when writing new value of current in status object.